Feature: close window / workspace from switcher

gabrielvincent opened the pull request:
Adds the ability to close windows or workspaces in the switcher by pressing `w`.

  ### Key Changes

  **Feature:**
  - Press `w` in the switcher to close the currently selected window or workspace
  - If the item failed to close (e.g., browser confirmation dialog), it gets focused for user interaction

  **Refactoring:**
  - **Extracted rendering logic from `open_switch` into dedicated `render_switch` function** (`windows-lib/src/switch/render.rs`)
    - Separates data collection from UI rendering
    - Enables reuse for refresh operations

  **New modules:**
  - `refresh.rs` - Handles switcher state refresh after item deletion
  - `render.rs` - Contains all GTK widget creation and rendering logic

  **Implementation details:**
  - Added `kill_client()` in exec-lib for closing windows
  - New transfer types: `CloseSwitchItem` and `RefreshSwitch`
  - Smart position-based selection after deletion in `next.rs`

  ### Testing
  - Open switcher with multiple windows/workspaces
  - Press `w` to close items
  - Verify correct next item selection
  - Test with apps that show confirmation dialogs (browsers)

gabrielvincent committed:
implement window closing on 'w'

H3rmt commented:
Looks interesting. 

I might reuse this functionality to add similar functionality to the overview mode. 

Like Ctrl + mouse click to close.

H3rmt commented on crates/windows-lib/src/switch/close.rs:
```diff
@@ -54,3 +57,52 @@ pub fn close_switch(data: &mut WindowsSwitchData, switch: bool) {
+fn kill_switch_client(data: &WindowsSwitchData) {
+    if let Some(id) = data.active.client {
+        let addr = to_client_address(id);
+        let _ = kill_client(addr);
```
please propagate the error upwards and handle it at some point (using .warn_details("---"), which logs the error)

H3rmt commented on crates/windows-lib/src/switch/create.rs:
```diff
@@ -144,6 +144,12 @@ fn handle_key(key: Key, event_sender: &Sender<TransferType>) -> Propagation {
                 .warn_details("unable to send");
             Propagation::Stop
         }
+        Key::w => {
```
hmm, any specific reason for w?

I personaly would use q. 
Maybe make configurable in the future.

The next version will include some config changes so I will add this in myself together with the other changes.

gabrielvincent replied:
That's the first key that came to my mind as I use ctrl+w to close tabs on the browser, on Nautilus, on Thunderbird, etc, and also I use Super + w to close windows on Hyprland, so it was a natural choice for me. I understand not everybody may have the same mental model. I don't oppose Q instead of W for a default. I think having the ability to configure this would be ideal.

H3rmt commented on crates/windows-lib/src/switch/close.rs:
```diff
@@ -54,3 +57,52 @@ pub fn close_switch(data: &mut WindowsSwitchData, switch: bool) {
+
+    for client_id in clients_to_kill {
+        let addr = to_client_address(client_id);
+        let _ = kill_client(addr);
```
same here

H3rmt commented on crates/windows-lib/src/switch/close.rs:
```diff
@@ -54,3 +57,52 @@ pub fn close_switch(data: &mut WindowsSwitchData, switch: bool) {
+    }
+
+    let sender = event_sender.clone();
+    glib::timeout_add_local(Duration::from_millis(50), move || {
```
any reason why you implemented this like this?

I mean sending the event from gui to kill the window and then (after a delay) sending another event to handle updating the switch ui.
Why not directly update the ui after receiving and executing the kill.

gabrielvincent replied:
I tried a different approach that uses a Hyprland event listener to detect when a window is closed. The problem with that is that when an application interrupts the closing of a window, the hyprland event is never sent. From the user's perspective, it's like their `w` command didn't have any effect as the item in the switcher won't disappear. To circumvent that, I added a small timeout after a `w` command to focus the window in case it hasn't been closed. Chances are that if it didn't close after 50ms, it means that something interrupted the close action, so we focus the window to bring it to the user's attention.

Since I already had to implement that workaround anyway, I decided to simplify the implementation. Everytime the user closes a window from the switcher, we wait 50ms, grab the current state, re-collect Hyprland data, compare current state with new state and derive the result of the close command from that comparison.

The reason why I'm sending an event instead of calling `refresh_after_close_item` directly is because when I try to pass `data: &WindowsSwitchData` to it from within the timeout callback, the borrow checker complains about `data: &WindowsSwitchData`'s lifetime:
```
borrowed data escapes outside of function
   `data` escapes the function body here [E0521]
2. borrowed data escapes outside of function
   argument requires that `'1` must outlive `'static` [E0521]
```

I thought this approach would be cleaner than messing with lifetimes. To be honest, this is the very first Rust code base I work on, so I'm still not very used to the memory management paradigm. If you think of a more idiomatic approach to this, please, let me know and I can learn it to make sure we have the best implementation.

H3rmt commented on crates/windows-lib/src/next.rs:
```diff
@@ -208,21 +218,20 @@ pub fn find_next_workspace(
-    let current = workspaces.iter().position(|w| w.0 == active).unwrap_or(0);
-    #[allow(clippy::cast_possible_wrap, clippy::cast_sign_loss)]
-    workspaces
+    let current_pos = enabled_workspaces
```
I will trust you on these changes

H3rmt commented on crates/windows-lib/src/switch/close.rs:
```diff
@@ -54,3 +57,52 @@ pub fn close_switch(data: &mut WindowsSwitchData, switch: bool) {
     }
 }
+
+pub fn close_switch_item(data: &WindowsSwitchData, event_sender: &Sender<TransferType>) {
```
please create a new file like `close_item` to contain this new code, as it doesnt handle closing the switch ui window

gabrielvincent committed:
bubble error result up and handle it with a warning log

additionally, extract functions related to the closing of a switch item
to its own file

H3rmt commented:
Sry didnt have time to test these features until now

H3rmt commented:
I will pull after I finish https://github.com/H3rmt/hyprshell/issues/367#issuecomment-3487686635.

fixing (adding) special workspaces is more complex than I thought.

gabrielvincent commented:
> I will pull after I finish [#367 (comment)](https://github.com/H3rmt/hyprshell/issues/367#issuecomment-3487686635).

Let me know if I can help with anything.

gabrielvincent committed:
fix conflicts with main branch

gabrielvincent committed:
increase timeout

H3rmt committed:
Merge branch 'hyprshell' into feature/close-switch-items
